Realm/ObjectStore/impl/background_collection.cpp

```diff
@@ -0,0 +1,832 @@
+    }
+}
+
+void BackgroundCollection::prepare_handover()
+{
+    REALM_ASSERT(m_sg);
+    m_sg_version = m_sg->get_version_of_current_transaction();
+    do_prepare_handover(*m_sg);
+}
+
+bool BackgroundCollection::deliver(SharedGroup& sg, std::exception_ptr err)
+{
+    if (!is_for_current_thread()) {
+        return false;
+    }
+
+    if (err) {
+        m_error = err;
+        return have_callbacks();
+    }
+
+    auto realm_sg_version = sg.get_version_of_current_transaction();
+    if (version() != realm_sg_version) {
+        // Realm version can be newer if a commit was made on our thread or the
+        // user manually called refresh(), or older if a commit was made on a
+        // different thread and we ran *really* fast in between the check for
+        // if the shared group has changed and when we pick up async results
+        return false;
+    }
+
+    bool should_call_callbacks = do_deliver(sg);
+    m_changes_to_deliver = std::move(m_accumulated_changes);
+
+    // fixup modifications to be source rows rather than dest rows
+    m_changes_to_deliver.modifications.erase_at(m_changes_to_deliver.insertions);
+    m_changes_to_deliver.modifications.shift_for_insert_at(m_changes_to_deliver.deletions);
+
+    return should_call_callbacks && have_callbacks();
+}
+
+void BackgroundCollection::call_callbacks()
+{
+    while (auto fn = next_callback()) {
+        fn(m_changes_to_deliver, m_error);
+    }
+
+    if (m_error) {
+        // Remove all the callbacks as we never need to call anything ever again
+        // after delivering an error
+        std::lock_guard<std::mutex> callback_lock(m_callback_mutex);
+        m_callbacks.clear();
+    }
+}
+
+CollectionChangeCallback BackgroundCollection::next_callback()
+{
+    std::lock_guard<std::mutex> callback_lock(m_callback_mutex);
+
+    for (++m_callback_index; m_callback_index < m_callbacks.size(); ++m_callback_index) {
+        auto& callback = m_callbacks[m_callback_index];
+        if (!m_error && callback.initial_delivered && m_changes_to_deliver.empty()) {
+            continue;
+        }
+        callback.initial_delivered = true;
+        return callback.fn;
+    }
+
+    m_callback_index = npos;
+    return nullptr;
+}
+
+void BackgroundCollection::attach_to(SharedGroup& sg)
+{
+    REALM_ASSERT(!m_sg);
+
+    m_sg = &sg;
+    do_attach_to(sg);
+}
+
+void BackgroundCollection::detach()
+{
+    REALM_ASSERT(m_sg);
+    do_detach_from(*m_sg);
+    m_sg = nullptr;
+}
+
+CollectionChangeBuilder::CollectionChangeBuilder(IndexSet deletions,
+                                                 IndexSet insertions,
+                                                 IndexSet modifications,
+                                                 std::vector<Move> moves)
+{
+    this->deletions = std::move(deletions);
+    this->insertions = std::move(insertions);
+    this->modifications = std::move(modifications);
+    this->moves = std::move(moves);
```

Comment:
Any reason to initialize these in the constructor body rather than as part of the initialization list?